### PR TITLE
integrate lighting effect

### PR DESCRIPTION
**Pull request by jianhuang01**
For #2096 

#### Change List
- integrate effects
- project point light source

### CODE REVIEW

File: modules/core/src/lib/effect-manager.js [hunk 1]
```diff
@@ -1,6 +1,24 @@
 export default class EffectManager {
   constructor() {
     this.effects = [];
+    this._needsRedraw = 'Initial render';
+  }
+
+  setProps(props) {
+    if ('effects' in props) {
+      if (this.effects !== props.effects) {
```
Comment (Pessimistress) [inline modules/core/src/lib/effect-manager.js]: Do we need deep comparison here?

File: modules/core/src/passes/draw-layers-pass.js [hunk 2]
```diff
@@ -33,4 +42,36 @@ export default class DrawLayersPass extends LayersPass {
     }
     return parameters;
   }
+
+  // Pre-project point light positions here
+  getEffectParameters(viewport, coordinateSystem, coordinateOrigin, effectProps) {
+    if (effectProps && effectProps.lightSources) {
+      const {pointLights} = effectProps.lightSources;
```
Comment (Pessimistress) [inline modules/core/src/passes/draw-layers-pass.js]: Shouldn't this be part of the LightEffect class?

File: modules/core/src/lib/effect-manager.js (hunk 1 above)
Comment (jianhuang01) [reply modules/core/src/lib/effect-manager.js]: good suggestion!

File: modules/core/src/passes/draw-layers-pass.js (hunk 2 above)
Comment (jianhuang01) [reply modules/core/src/passes/draw-layers-pass.js]: no, point light's projection is tight to viewport and layer's coordinateSystem, effect and effectManager doesn't know this information. In the layers-pass where we have these viewport and layer information we shouldn't see the effect objects, only the final props(uniform, texture, etc.) from effects. In long term once we have a unified world space we won't need this projection code.

File: modules/core/src/lib/deck.js [hunk 3]
```diff
@@ -218,6 +218,10 @@ export default class Deck {
       this.layerManager.setProps(newProps);
     }
 
+    if (this.effectManager) {
+      this.effectManager.setProps(newProps);
```
Comment (ibgreen) [inline modules/core/src/lib/deck.js]: Nit: Also call finalize on effect manager in Deck.finalize?

File: modules/core/src/lib/deck.js [hunk 4]
```diff
@@ -250,9 +254,15 @@ export default class Deck {
 
     const viewManagerNeedsRedraw = this.viewManager.needsRedraw({clearRedrawFlags});
     const layerManagerNeedsRedraw = this.layerManager.needsRedraw({clearRedrawFlags});
+    const effectManagerNeedsRedraw = this.effectManager.needsRedraw({clearRedrawFlags});
```
Comment (ibgreen) [inline modules/core/src/lib/deck.js]: Unrelated: animation loop can also need redraw... @tsherif

File: modules/core/src/lib/deck.js [hunk 5]
```diff
@@ -523,12 +533,9 @@ export default class Deck {
...
     this.effectManager = new EffectManager();
 
-    this.deckRenderer = new DeckRenderer(gl, {
-      layerManager: this.layerManager,
-      effectManager: this.effectManager
-    });
+    this.deckRenderer = new DeckRenderer(gl);
 
-    this.deckPicker = new DeckPicker(gl, {layerManager: this.layerManager});
+    this.deckPicker = new DeckPicker(gl);
```
Comment (ibgreen) [inline modules/core/src/lib/deck.js]: +1 nice decoupling!

File: modules/core/src/passes/draw-layers-pass.js [hunk 6]
```diff
@@ -2,9 +2,12 @@ import LayersPass from './layers-pass';
 
 export default class DrawLayersPass extends LayersPass {
   // PRIVATE
-  getModuleParameters(layer, pixelRatio) {
+  getModuleParameters(layer, pixelRatio, effects, effectProps) {
     const moduleParameters = super.getModuleParameters(layer, pixelRatio);
-    Object.assign(moduleParameters, this.getObjectHighlightParameters(layer));
+    Object.assign(moduleParameters, this.getObjectHighlightParameters(layer), effectProps);
+    for (const effect of effects) {
```
Comment (ibgreen) [inline modules/core/src/passes/draw-layers-pass.js]: * At first glance, it seems a little strange that the DrawLayersPass has knowledge of other effects. I would expect e.g. the renderer to call something like `effectManager.extractGlobalSettingsFromEffects()` before calling the render passes.
* Even if we want to/need to do it here, I would still encapsulate the iteration of the effects inside the effectManager, e.g. `effectManager.getParametersForLayer()`

File: modules/core/src/passes/layers-pass.js [hunk 7]
```diff
@@ -19,7 +19,8 @@ export default class LayersPass extends Pass {
     pass = 'draw',
     redrawReason = '',
     customRender,
-    effects
+    effects,
```
Comment (ibgreen) [inline modules/core/src/passes/layers-pass.js]: Perhaps just pass down the `effectManager` depending on how we resolve previous comment.

File: modules/core/src/lib/deck.js (hunk 4 above)
Comment (Pessimistress) [reply modules/core/src/lib/deck.js]: This recurssively getting redraw flag thing is getting quite heavy. We might want to consider moving it to a push model.

**Pull request merged**